### PR TITLE
ANDROID-11179 detekt implementation

### PR DESCRIPTION
**Pull request by jmanriquehiberus**
### :tickets: Jira ticket
[ANDROID-11179](https://jira.tid.es/browse/ANDROID-11179)

### :goal_net: What's the goal?
_We need to implement detekt._

### :construction: How do we do it?
`mistica-android/detekt-baseline.xml`
Some custom code smells added so detekt can spot them (taken from android-messenger).

`mistica-android/build-tools/detekt/detekt.yml`
Custom definition of detekt's behaviour.

`mistica-android/build-tools/detekt.gradle`
Gradle file, applied from the project's **build.gradle** in order to override gradle's detekt.

I also fixed 57 detekt issues, most of them are fixed in atomic commits in order to ease the review. Still, there are 39 issues to be addressed which are suppressed by now. I will create now some jira tickets to tackle them.

### :test_tube: How can I test this?
Run the command `./gradlew detekt` in order to run the task, modify **detekt.yml** and see that those new changes affect the output of the task.

### CODE REVIEW

**Comment by jmanriquehiberus** [comment]
Should we include detekt in any of our GAs?

File: build-tools/detekt.gradle [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+detekt {
+    buildUponDefaultConfig = true
+    config = files("$projectDir/build-tools/detekt/detekt.yml")
+    input = files("$projectDir")
+}
```
Comment (juangardi21) [inline build-tools/detekt.gradle:6]: You could exclude some folders like:
```
    source.from(fileTree(rootProject.projectDir) {
        exclude("**/.gradle/")
        exclude("**/build/")
        exclude("**/tmp/**")
    })
```

**Review by juangardi21 — APPROVED** [review]
LGTM, just one suggestion 👍🏼

File: build-tools/detekt.gradle (hunk 1 above)
Comment (jmanriquehiberus) [reply build-tools/detekt.gradle:12]: True, just did it, thanks ☺️

File: build-tools/detekt.gradle [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
+detekt {
```
Comment (dpastor) [inline build-tools/detekt.gradle:2]: No files to adapt in mistica after adding detekt? Did you check this is executed with the ./gradlew check command? (which is the one executed on preintegration GA)

Comment (jmanriquehiberus) [reply build-tools/detekt.gradle:2]: Yes, there are 96 issues to address after running detekt. 

I am sorry but could not find any preintegration GA in mistica, should there be or should we include detekt in any GA? 🤔

Comment (dpastor) [reply build-tools/detekt.gradle:2]: It's the tests.yml file, the checks which are run on each PR commit, detekt should run as part of them.

Normally, when invoking "check" gradle target (executed on that GA), detekt is also executed with just applying the plugin, but please check it just to ensure it.

Comment (dpastor) [reply build-tools/detekt.gradle:2]: In order to add detekt we need to also fix the current issues repository has (or ignoring them someway to do that work afterwards), if it's too complicated, please check with the team how to proceed on next weekly or by chat (I'm on vacations next week).

Comment (jmanriquehiberus) [reply build-tools/detekt.gradle:2]: Okay. By now, I have addressed 53 issues and still have 43 to be fixed (most of them related to long functions or too complex functions). I will address simpler issues alone and will keep potentially troublesome ones to be tackled along with the team.

File: build.gradle [hunk 3]
```diff
@@ -41,10 +48,50 @@ allprojects {
         version = System.getProperty("LIBRARY_VERSION") ?: "undefined"
     }
 
+    tasks.withType(io.gitlab.arturbosch.detekt.Detekt).all {
+        exclude("build/")
+        exclude("**/build/**")
+        exclude("**/resources/**")
+        exclude("**/tmp/**")
+    }
```
Comment (jeprubio) [inline build.gradle:56]: these are also in detekt.gradle ¿are they needed in both places? ¿Perhaps if this is here the ones from detekt.gradle are not needed?

File: detekt-baseline.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" ?>
+<SmellBaseline>
+  <ManuallySuppressedIssues></ManuallySuppressedIssues>
+  <CurrentIssues>
+    <ID>WildcardImport:protect_android_task.gradle.kts$import java.io.*</ID>
+    <ID>WildcardImport:protect_android_task.gradle.kts$import java.lang.*</ID>
+  </CurrentIssues>
+</SmellBaseline>
```
Comment (jeprubio) [inline detekt-baseline.xml:8]: I think we could either fix these imports or use @Supress* in that files in order not to need this baseline. I fear we'll fix this at some point and forget to remove the baseline.

**Review by jeprubio — APPROVED** [review]
👍

File: build.gradle (hunk 3 above)
Comment (jmanriquehiberus) [reply build.gradle]: True, just committed a code cleanup. Thanks!

File: detekt-baseline.xml (hunk 4 above)
Comment (jmanriquehiberus) [reply detekt-baseline.xml]: I did not have to fix any issues here, it was my bad. Thanks for spotting it.

Thinking it twice, it makes no sense to have an empty detekt-baseline.xml file for now, so I am deleting it.

**Review by juangardi21 — APPROVED** [review]
Awesome

**Comment by nimeacuerdo** [comment]
Breaking changes that require a new major release? Any ticket created to tackle pending issues?

**Comment by jmanriquehiberus** [comment]
> Breaking changes that require a new major release? if so I would expect updates in https://github.com/Telefonica/mistica-android/blob/jmanrique/ANDROID-11179-enable-detekt-in-mistica-android/UPGRADING.md
> 
> Any ticket created to tackle pending issues?

I have just created these 3 tickets to tackle pending issues.
- https://jira.tid.es/browse/ANDROID-15230
- https://jira.tid.es/browse/ANDROID-15231
- https://jira.tid.es/browse/ANDROID-15232

I have a doubt about this being a major release. I would say this is a minor one, since there are no breaking changes (I think), it is just adding new functionality (detekt) which is backward compatible. What am I missing?

**Comment by nimeacuerdo** [comment]
> > Breaking changes that require a new major release? if so I would expect updates in https://github.com/Telefonica/mistica-android/blob/jmanrique/ANDROID-11179-enable-detekt-in-mistica-android/UPGRADING.md
> > Any ticket created to tackle pending issues?
> 
> I have just created these 3 tickets to tackle pending issues.
> 
> * https://jira.tid.es/browse/ANDROID-15230
> * https://jira.tid.es/browse/ANDROID-15231
> * https://jira.tid.es/browse/ANDROID-15232
> 
> I have a doubt about this being a major release. I would say this is a minor one, since there are no breaking changes (I think), it is just adding new functionality (detekt) which is backward compatible. What am I missing?

what about this https://github.com/Telefonica/mistica-android/pull/386/files#diff-16c7f6781a03c2127c9cfd4b50a25cd8ad71e464e4848fccbcc9a823e44f73bdL222-R222?

File: catalog/src/main/java/com/telefonica/mistica/catalog/ui/compose/components/Buttons.kt [hunk 5]
```diff
@@ -50,6 +50,7 @@ fun Buttons() {
 }
 
 @Composable
+@Suppress("LongMethod")
```
Comment (jeslat) [inline catalog/src/main/java/com/telefonica/mistica/catalog/ui/compose/components/Buttons.kt:53]: I would include an ignore to not take into account the composables as we do in [platform](https://github.com/Telefonica/novum-platform-android/blob/main/build-tools/detekt/detekt.yml#L6-L7):
```
  LongMethod:
    ignoreAnnotated: 'Composable'
```

In general, most of the composables will throw the long method error

**Comment by jeslat** [comment]
> > > > Breaking changes that require a new major release? if so I would expect updates in https://github.com/Telefonica/mistica-android/blob/jmanrique/ANDROID-11179-enable-detekt-in-mistica-android/UPGRADING.md
> > > > Any ticket created to tackle pending issues?
> > > 
> > > 
> > > I have just created these 3 tickets to tackle pending issues.
> > > 
> > > * https://jira.tid.es/browse/ANDROID-15230
> > > * https://jira.tid.es/browse/ANDROID-15231
> > > * https://jira.tid.es/browse/ANDROID-15232
> > > 
> > > I have a doubt about this being a major release. I would say this is a minor one, since there are no breaking changes (I think), it is just adding new functionality (detekt) which is backward compatible. What am I missing?
> > 
> > 
> > Thanks for creating those tickets!
> > what about this https://github.com/Telefonica/mistica-android/pull/386/files#diff-16c7f6781a03c2127c9cfd4b50a25cd8ad71e464e4848fccbcc9a823e44f73bdL222-R222?
> 
> Sorry, thanks!! I did not know that was a breaking change, I am adding it to https://github.com/Telefonica/mistica-android/blob/jmanrique/ANDROID-11179-enable-detekt-in-mistica-android/UPGRADING.md as you mentioned.

@jmanriquehiberus as it's explained in that document, now the breaking changes are explained in the release notes of the new version, in this case `15.0.0`. You have an example here https://github.com/Telefonica/mistica-android/releases/tag/14.0.0

File: catalog/src/main/java/com/telefonica/mistica/catalog/ui/compose/components/Buttons.kt (hunk 5 above)
Comment (jmanriquehiberus) [reply catalog/src/main/java/com/telefonica/mistica/catalog/ui/compose/components/Buttons.kt]: Great, I have just included the ignore you mentioned and there are only 2 LongMethod flagged non composable functions. Thank you!!!

**Review by jeslat — APPROVED** [review]
Thanks, good job!

**Pull request merged**